File: Безопасная разработка ПО/4_лаба/cpp/main.cpp
#include <iostream>
#include <string>
#include <sstream>
#include "stack.h"
#include "queue.h"
#include "tree.h"

void showHelp()
{
    std::cout << "s.push <value>   - добавить элемент в стек\n";
    std::cout << "s.pop            - удалить верхний элемент из стека\n";
    std::cout << "s.peek           - посмотреть верхний элемент стека\n";
    std::cout << "s.print          - вывести стек\n";
    std::cout << "q.push <value>   - добавить элемент в очередь\n";
    std::cout << "q.pop            - удалить первый элемент из очереди\n";
    std::cout << "q.front          - посмотреть первый элемент очереди\n";
    std::cout << "q.print          - вывести очередь\n";
    std::cout << "t.push <value>   - добавить элемент в дерево\n";
    std::cout << "t.pop            - удалить элемент из дерева\n";
    std::cout << "t.print          - вывести дерево (обход)\n";
    std::cout << "t.height         - показать высоту дерева\n";
    std::cout << "help             - показать список команд\n";
    std::cout << "exit             - выход из программы\n\n";   
}

int main() {
    Stack s;
    Queue q;
    Tree t;
    std::string command;

    std::cout << "help для списка команд.\n";

    while (true) {
        std::cout << "> ";
        std::getline(std::cin, command);

        std::istringstream iss(command);
        std::string cmd;
        iss >> cmd;

        if (cmd == "exit") {
            std::cout << "Выход из программы.\n";
            break;
        }
        else if (cmd == "help") {
            showHelp();
        }
        else if (cmd == "s.push") {
            int value;
            if (iss >> value) {
                s.push(value);
                std::cout << "Элемент " << value << " добавлен в стек.\n";
            }
            else {
                std::cout << "Ошибка: требуется значение.\n";
            }
        }
        else if (cmd == "s.pop") {
            s.pop();
        }
        else if (cmd == "s.peek") {
            std::cout << "Верхний элемент: " << s.peek() << "\n";
        }
        else if (cmd == "s.print") {
            s.print();
        }
        else if (cmd == "q.push") {
            int value;
            if (iss >> value) {
                q.push(value);
                std::cout << "Элемент " << value << " добавлен в очередь.\n";
            }
            else {
                std::cout << "Ошибка: требуется значение.\n";
            }
        }
        else if (cmd == "q.pop") {
            q.pop();
        }
        else if (cmd == "q.front") {
            std::cout << "Первый элемент: " << q.front() << "\n";
        }
        else if (cmd == "q.print") {
            q.print();
        }
        else if (cmd == "t.push") {
            int value;
            if (iss >> value) {
                t.push(value);
                std::cout << "Элемент " << value << " добавлен в дерево.\n";
            }
            else {
                std::cout << "Ошибка: требуется значение.\n";
            }
        }
        else if (cmd == "t.pop") {
            t.pop();
        }
        else if (cmd == "t.print") {
            t.print();
        }
        else if (cmd == "t.height") {
            std::cout << "Высота дерева: " << t.height() << "\n";
        }
        else if (!cmd.empty()) {
            std::cout << "Неизвестная команда: '" << cmd << "'. Введите 'help'\n";
        }
    }

    return 0;
}

File: Безопасная разработка ПО/4_лаба/python/main.py
from stack import Stack

def show_help():
    print("Команды:")
    print("s.push <value>   - добавить элемент в стек")
    print("s.pop            - удалить верхний элемент из стека")
    print("s.peek           - посмотреть верхний элемент стека")
    print("s.print          - вывести содержимое стека")
    print("s.size           - показать размер стека")
    print("help             - показать список команд")
    print("exit             - выход из программы")
    print()

def main():
    stack = Stack()
    print("help для списка команд.")

    while True:
        print("> ", end="")
        command = input().strip()

        parts = command.split()
        if not parts:
            continue

        cmd = parts[0]

        if cmd == "exit":
            print("Выход из программы.")
            break

        elif cmd == "help":
            show_help()

        elif cmd == "s.push":
            if len(parts) < 2:
                print("Ошибка: требуется значение")
            else:
                value = int(parts[1])
                stack.push(value)

        elif cmd == "s.pop":
            popped_value = stack.pop()
            print(f"Удаленный элемент: {popped_value}")

        elif cmd == "s.peek":
            try:
                top_value = stack.peek()
                print(f"Верхний элемент стека: {top_value}")
            except IndexError:
                print("Ошибка: стек пуст")

        elif cmd == "s.print":
            stack.print_stack()

        elif cmd == "s.size":
            print(f"Размер стека: {stack.size()}")

        else:
            print(f"Неизвестная команда: '{cmd}'. help для списка команд")

if __name__ == "__main__":
    main()
